pkg/internal/cluster/providers/provider.go

```diff
@@ -0,0 +1,40 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package providers
+
+import (
+	"sigs.k8s.io/kind/pkg/cluster/nodes"
+
+	"sigs.k8s.io/kind/pkg/internal/apis/config"
+)
+
+// Provider represents a provider of cluster / node infrastructure
+type Provider interface {
```

Comment:
nice

Reply:
once everything is retargeted on this I'm going to start using a mock one in a few places and then podman, ignite etc. :D

between this and the node interface, almost all of the rest of the code is much higher level and no longer needs to think in terms of `docker exec ...` etc. and depends on a much narrower set of methods.

the docker provider has quite a lot of internal code that used to be spread over several packages.

the reusable bits I've been moving into a shared library though (like getting an open port on the host) to ease implementing podman etc.